merkletree/merkletree.go

```diff
@@ -478,26 +478,49 @@ node [fontname=Monospace,fontsize=10,shape=box]
 
 // DumpTree outputs a list of all the key value in hex. Notice that this will
 // output the full tree, which is not needed to reconstruct the Tree. To
-// reconstruct the tree can be done from the output of DumpClaims funtion.
-// The difference between DumpTree and DumpClaims is that with DumpTree the
-// size of the output will be almost the double but to recover the tree will
-// not need to compute the Tree, while with DumpClaims will require to compute
-// the Tree (with the computational cost of each hash)
+// reconstruct the tree can be done from the output of DumpClaims funtion.  The
+// difference between DumpTree and DumpClaims is that with DumpTree the size of
+// the output will be almost the double (in raw bytes, but with the current
+// implementation DumpTree output is even smaller than DumpClaims output size, but because DumpTree stores in
+// binary while DumpClaims stores in hex) but to recover the tree will not need
+// to compute the Tree, while with DumpClaims will require to compute the Tree
+// (with the computational cost of each hash)
 func (mt *MerkleTree) DumpTree(w io.Writer, rootKey *Hash) error {
 	err := mt.Walk(rootKey, func(n *Node) {
 		if n.Type != NodeTypeEmpty {
-			fmt.Fprintf(w, "%v:%v,", n.Key().Hex(), common3.HexEncode(n.Value()))
+			serializeKV(w, n.Key().Bytes(), n.Value())
 		}
 	})
 
 	if rootKey == nil {
 		rootKey = mt.RootKey()
 	}
-	fmt.Fprintf(w, "%v:%v,", common3.HexEncode(rootNodeValue), rootKey.Hex())
+	serializeKV(w, rootNodeValue, rootKey.Bytes())
 
 	return err
 }
 
+func serializeKV(w io.Writer, k, v []byte) {
+	kH := byte(len(k))
+	vH := common3.Uint16ToBytes(uint16(len(v)))
+	w.Write([]byte{kH})
+	w.Write(vH)
+	w.Write(k)
+	w.Write(v)
+}
+
+func deserializeKV(b []byte) ([]byte, []byte, error) {
```

Comment:
It's a bit weird that this function parses the header where `ImportTree` has already parsed the header.  Maybe it'd be better if this function takes the writer and returns a special error on EOF.